### ISSUE TITLE
Unexpected Band Structure Results and Extremely Low Hamiltonian MAE (9.13e-07) in HamGNN Training for CsPbI3 MD Snapshots

### ISSUE THREAD

**Issue by liee11-bit** [QuantumLab-ZY/HamGNN#70]
Deer Developer of HamGNN
I'm a user of HamGNN to build a CsPbI3 model using 100 MD snapshots as training data. The training process completes without errors, but the predicted band structure shows incorrect results, despite achieving an unusually low Hamiltonian MAE of 9.13e-07.
**Workflow Summary:**
**1DFT Data Preparation:**
Ran mpirun -np 32 abacus_postprocess → data-H0R-sparse_SPIN0.csr and data-S0R-sparse_SPIN0.csr
Ran mpirun -np 24 abacus → data-HR-sparse_SPIN0.csr, data-SR-sparse_SPIN0.csr, and running_scf.log
Packaged 100 structures into graph_data.npz
Key Parameters:
''RADIUS_SCALE_FACTOR = 1.8
SKIP_DFT_HAMILTONIAN = True
SOC_ENABLED = False
NUM_PROCESSES = 2''
**2. Training Process:**

- First training: HamGNN2.0 --config config.yaml
 ''dataset_params:
  batch_size: 1
  split_file: null
  test_ratio: 0.1
  train_ratio: 0.7
  val_ratio: 0.2
  graph_data_path: /share/home/liyuting/TEST/cspbi3_ham/work_dir/dataset/graph/new # Directory wh
ere graph_data.npz is located

losses_metrics:
  losses:
  - loss_weight: 1.0
    metric: mae
    prediction: hamiltonian
    target: hamiltonian
  #- loss_weight: 1.0
  #  metric: mae
  #  prediction: band_gap
  #  target: band_gap
  # - loss_weight: 0.2721
  #   metric: mae
  #   prediction: band_energy
  #   target: band_energy
  - loss_weight: 1.0
    metric: mae
    prediction: overlap
    target: overlap
  #- loss_weight: 1.0
  #  metric: mae
  #  prediction: peak
  #  target: peak
  # - loss_weight: 0.0
  #   metric: mae
  #   prediction: hamiltonian_imag
  #   target: hamiltonian_imag
  #- loss_weight: 0.0001
  #  metric: abs_mae
  #  prediction: wavefunction
  #  target: wavefunction
  metrics:
  - metric: mae
    prediction: hamiltonian
    target: hamiltonian
  #- metric: mae
  #  prediction: band_gap
  #  target: band_gap
  #- metric: mae
  #  prediction: peak
  #  target: peak
  - metric: mae
    prediction: overlap
    target: overlap
  # - metric: mae
  #   prediction: hamiltonian_imag
  #   target: hamiltonian_imag
  #- metric: mae
  #  prediction: hamiltonian_imag
  #  target: hamiltonian_imag
  # - metric: mae
  #   prediction: band_energy
  #   target: band_energy
  #- metric: abs_mae
  #  prediction: wavefunction
  #  target: wavefunction

# Generally, the optim_params module only needs to set the initial learning rate (lr)
optim_params:
  lr: 0.01    #First training:0.01;Second training:0.0001
  lr_decay: 0.5
  lr_patience: 5
  gradient_clip_val: 0.0
  max_epochs: 3000
  min_epochs: 100
  stop_patience: 30

output_nets:
  output_module: HamGNN_out
  HamGNN_out:
    ham_only: false # true: Only the Hamiltonian H is computed; 'false': Fit both H and S
    ham_type: abacus # openmx: fit openmx Hamiltonian; abacus: fit abacus Hamiltonian
    nao_max: 27 # The maximum number of atomic orbitals in the data set, which can be 14, 19 or 2
7
    add_H0: True # Generally true, the complete Hamiltonian is predicted as the sum of H_scf plus
 H_nonscf (H0)
    symmetrize: True # if set to true, the Hermitian symmetry constraint is imposed on the Hamilt
onian
    calculate_band_energy: False # Whether to calculate the energy bands to train the model
    num_k: 5 # When calculating the energy bands, the number of K points to use
    band_num_control: 8 # `dict`: controls how many orbitals are considered for each atom in ener
gy bands; `int`: [vbm-num, vbm+num]; `null`: all bands
    k_path: null # `auto`: Automatically determine the k-point path; `null`: random k-point path;
 `list`: list of k-point paths provided by the user
    soc_switch: False # if true, fit the SOC Hamiltonian
    nonlinearity_type: gate # norm or gate
    # spin constrained
    spin_constrained: False
    collinear_spin: False
    minMagneticMoment: 0.5

profiler_params:
  progress_bar_refresh_rat: 1
  train_dir: /share/home/liyuting/TEST/cspbi3_ham/work_dir/model_dir #The folder for saving train
ing information and prediction results. This directory can be read by tensorboard to monitor the 
training process.

representation_nets:
  # Network parameters usually do not need to be changed.
  HamGNN_pre:
    cutoff: 26.0
    cutoff_func: cos
    edge_sh_normalization: component
    edge_sh_normalize: true
    ######## Irreps set 1 (crystal): ################
    irreps_edge_sh: 0e + 1o + 2e + 3o + 4e + 5o
    irreps_node_features: 48x0e+48x0o+24x1o+12x1e+8x2o+15x2e+10x3o+5x3e
    num_layers: 3
    num_radial: 48
    num_types: 96
    rbf_func: bessel
    set_features: true
    radial_MLP: [64, 64]
    use_corr_prod: False
    correlation: 2
    num_hidden_features: 16
    use_kan: False
    radius_scale: 1.01
    build_internal_graph: False

setup:
  GNN_Net: HamGNNpre
  accelerator: null
  ignore_warnings: true
  checkpoint_path: null # Path to the model weights file
  load_from_checkpoint: False
  resume: False
  num_gpus: [0] # null: use cpu; [i]: use the ith GPU device
  precision: 32
  property: hamiltonian
  stage: fit # fit: training; test: inference
"

**Comment by liee11-bit** [comment]
Dear Developer of HamGNN
I refer to the content in the comment <https://github.com/QuantumLab-ZY/HamGNN/issues/47> and use the previous version of the graph script to realize the packaging of SOC Hamiltonian quantities. I have a new question about training. After I already have a model, if I want to add a structure to it, should I fine-tune the model directly on the basis of the original model or should I re-train it? I look forward to your reply.
Best
Yuting Li

**Comment by QuantumLab-ZY** [comment]
If the new structures you added are similar to those in your original training dataset, you can fine-tune the model based on your existing one.

Additionally, if you plan to use HamGNN to train SOC Hamiltonians generated by Abacus, I recommend reinstalling HamGNN v2.1 from GitHub. I recently fixed a bug related to training SOC Hamiltonians from Abacus in the latest release.

**Comment by liee11-bit** [comment]
Dear Developer of HamGNN
Thank you for your reply. I'll give it a try! Additionally, I have a new problem. The model I originally trained on small structures （40 ATOMS CsPbI3）does not perform well when predicting larger supercell structures（320 ATOMS CsPbI3）.
 <img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/1b8beba9-02e1-41a5-992a-9e6544fda2dd" />，Is it possible to adjust the model by adding a larger super cell structure as a training set?
Looking forward to your suggestions！
Best
Yuting Li

**Comment by QuantumLab-ZY** [comment]
You need to first check whether the predicted band structures for the training set are correct.

**Comment by QuantumLab-ZY** [comment]
It might be that the CsPbI3 supercell containing 320 atoms has a significant difference from the structures in your training set, not because it is a supercell. Or you might not have correctly packaged the graph_data.npz file for this structure, considering that some mistakes often occur in the packaging process. How about the MAE of the Hamiltonian for this structure?

**Comment by liee11-bit** [comment]
Dear Developer of HamGNN
Thank you for your previous assistance, which has given me some understanding of Hamgnn! Now, if I want to train a Hamgnn large model for perovskite structures (which include inorganic and organic-inorganic hybrid perovskites containing multiple different elements), how should I construct the training dataset? Could you offer some advice on selecting the training set? I look forward to your response.
Best
Yuting Li

**Comment by QuantumLab-ZY** [comment]
At the moment, we do not have an active-learning pipeline implemented for Hamiltonian matrix training. The recommended approach is to collect a comprehensive dataset of perovskite structures and compute their corresponding Hamiltonian matrices to serve as the training set.

If your dataset becomes very large, you can split it into several smaller subsets and train HamGNN on them sequentially. This can help manage memory usage and training efficiency.

**Comment by liee11-bit** [comment]
Thanks for your help, this issue has been resolved. I aim to train a model capable of predicting the Hamiltonian matrices of various perovskites across different phases and temperatures. Do you have any recommendations regarding the selection of training data? How many atomic-scale systems would be optimal for the training set and how many systems should be selected for each temperature? (The predicted structures typically contain around 300 atoms.)

**Comment by QuantumLab-ZY** [comment]
It is recommended to collect over 10,000 perovskite structures encompassing diverse types and compositions. Each structure should contain no more than approximately 50 atoms to prevent out of the memory error during the training process.
